### DISCUSSION TITLE
GraalVM flavor

### DISCUSSION THREAD

**Discussion by tomerr90** [paketo-buildpacks/feedback#278] [Java Team]
Im a bit confused here, Im also new to buildpacks so sorry if its a stupid question.
There only seems to be one type of buildpack for graalvm https://console.cloud.google.com/gcr/images/paketo-buildpacks/GLOBAL/graalvm
But I saw this https://github.com/paketo-buildpacks/rfcs/pull/294
Which seems to suggest a buildpack was added for Oracle GraalVM but I cant find it...
So my question is, what GraalVM is actually in the buildpack that I added above? Am I supposed to provide some config for it to "choose" between CE/EE/Oracle Graal?
Specifically Im looking for a buildpack for Java 11 CE Graal

**Comment by dmikusa** [comment]
The https://github.com/paketo-buildpacks/graalvm/ buildpack installs GraalVM CE. This one is the community offering under a standard OSS license with community support.

The https://github.com/paketo-buildpacks/oracle buildpack installs Oracle GraalVM. This one is the Oracle offering under Oracle's Free License with support available from Oracle.

There's also https://github.com/paketo-buildpacks/bellsoft-liberica which installs Bellsoft's NIK (which is also GraalVM). Based on the community offering but with support available from Bellsoft.

I'm not affiliated with any of those companies. I'm simplifying things greatly and I very likely missed one or more of the unique points and differentiators between these GraalVM offerings. I hope that's enough to point everyone to the right buildpacks, and y'all can do your own research on licenses, support, and that sort of stuff :) 

Totally reasonable question btw. We probably would have renamed the graalvm buildpack to graalvm-ce, but it's just really difficult to rename buildpacks cause it will break builds.

**Reply by tomerr90** [reply]
Well I guess you can create a new buildpack called graalvm-ce and put new versions there so I guess people will slowly move.
Either way, thank you!

**Comment by tomerr90** [comment]
One more question, so for the GraalVM CE, there are versions for different Java versions, i.e. there is version X of GraalVM for Java 11, and also for Java 17 for example.
Does each buildpack version contain all the Java versions available for that specific GraalVM version?

**Reply by dmikusa** [reply]
We ship the latest of each major version that's still supported by upstream, so right now that's 17 and 21. We should be adding 22 any time now.

You can always check what's included by looking at the release notes -> https://github.com/paketo-buildpacks/graalvm/releases

**Reply by tomerr90** [reply]
So
1. You dropped Java 11 support because its no longer maintained and supported right? Just 17/21?
2. If I would like to build the buildpack myself by basically copying the toml and adding back the 2 dependecies for Java 11 (jdk and the native image), do you think it will work?

**Reply by dmikusa** [reply]
1. Yes. Upstream doesn't support it any longer. That is when we drop old dependencies. Not just GraalVM, but all dependencies that are buildpack managed.
2. Yes. You can package it yourself. Just `git clone` and modify `buildpack.toml`. Then build & package.

Historically, it was a little convoluted to build your own Paketo buildpacks but we've been working on making this easier with tools.

https://github.com/paketo-buildpacks/libpak-tools?tab=readme-ov-file#libpak-tools-package-bundle

You can run `go install github.com/paketo-buildpacks/libpak-tools@latest` to install it and then `libpak-tools package bundle --buildpack-id <buildpack-id>` from the project root to get a working custom buildpack image.
